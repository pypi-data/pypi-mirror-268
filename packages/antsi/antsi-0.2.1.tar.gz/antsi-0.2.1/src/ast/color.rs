use std::{
    fmt::{Display, Formatter},
    str::FromStr,
};

/// The provided [`Color`] name was invalid
#[derive(Clone, Copy, Debug)]
pub struct InvalidColorError;

impl std::error::Error for InvalidColorError {}

impl Display for InvalidColorError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str("invalid color name")
    }
}

macro_rules! colors {
    (
        $( $( #[ $meta:meta ] )* $color:ident $fg:literal $bg:literal ( $names:pat ) ),* $(,)?
    ) => {
        /// Available standard ANSI colors
        #[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
        pub enum Color {
            $( $( #[$meta] )* $color, )*
        }

        impl Color {
            /// Convert to the foreground ANSI code
            pub fn foreground_code(&self) -> &'static str {
                match self {
                    $( Color::$color => stringify!($fg), )*
                }
            }

            /// Convert to the background ANSI code
            pub fn background_code(&self) -> &'static str {
                match self {
                    $( Color::$color => stringify!($bg), )*
                }
            }
        }

        impl FromStr for Color {
            type Err = InvalidColorError;

            fn from_str(name: &str) -> Result<Self, Self::Err> {
                Ok(match name.to_ascii_lowercase().as_str() {
                    $( $names => Color::$color, )*
                    _ => return Err(InvalidColorError),
                })
            }
        }
    };
}

colors! {
    Black   30 40 ("black"),
    Red     31 41 ("red"),
    Green   32 42 ("green"),
    Yellow  33 43 ("yellow"),
    Blue    34 44 ("blue"),
    Magenta 35 45 ("magenta"),
    Cyan    36 46 ("cyan"),
    White   37 47 ("white"),
    #[default]
    Default 39 49 ("default"),

    BrightBlack   90 100 ("bright-black"),
    BrightRed     91 101 ("bright-red"),
    BrightGreen   92 102 ("bright-green"),
    BrightYellow  93 103 ("bright-yellow"),
    BrightBlue    94 104 ("bright-blue"),
    BrightMagenta 95 105 ("bright-magenta"),
    BrightCyan    96 106 ("bright-cyan"),
    BrightWhite   97 107 ("bright-white"),
}
