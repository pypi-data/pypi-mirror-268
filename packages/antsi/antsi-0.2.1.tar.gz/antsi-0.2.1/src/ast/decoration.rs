use std::{
    fmt::{Display, Formatter},
    str::FromStr,
};

/// The provided [`Decoration`] name was invalid
#[derive(Clone, Copy, Debug)]
pub struct InvalidDecorationError;

impl std::error::Error for InvalidDecorationError {}

impl Display for InvalidDecorationError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str("invalid decoration name")
    }
}

macro_rules! decorations {
    (
        $( $decoration:ident $apply:literal $remove:literal ( $names:pat ) ),* $(,)?
    ) => {
        /// Available standard ANSI text decorations
        #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
        pub enum Decoration {
            $( $decoration, )*
        }

        impl Decoration {
            /// Convert to the ANSI code for applying the styling
            pub fn apply_code(&self) -> &'static str {
                match self {
                    $( Decoration::$decoration => stringify!($apply), )*
                }
            }

            /// Convert to the ANSI code for removing the styling
            pub fn remove_code(&self) -> &'static str {
                match self {
                    $( Decoration::$decoration => stringify!($remove), )*
                }
            }
        }

        impl FromStr for Decoration {
            type Err = InvalidDecorationError;

            fn from_str(name: &str) -> Result<Self, Self::Err> {
                Ok(match name.to_ascii_lowercase().as_str() {
                    $( $names => Decoration::$decoration, )*
                    _ => return Err(InvalidDecorationError),
                })
            }
        }
    };
}

decorations! {
    Bold          1 22 ("bold"),
    Dim           2 22 ("dim" | "faint"),
    Italic        3 23 ("italic"),
    Underline     4 24 ("underline"),
    SlowBlink     5 25 ("slow-blink" | "blink-slow"),
    FastBlink     6 25 ("fast-blink" | "blink-fast"),
    Invert        7 27 ("invert" | "reverse"),
    Hide          8 28 ("hide" | "conceal"),
    StrikeThrough 9 29 ("strikethrough" | "strike-through"),
}
