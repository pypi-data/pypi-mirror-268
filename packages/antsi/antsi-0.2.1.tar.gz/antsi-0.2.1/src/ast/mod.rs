mod color;
mod decoration;
mod style;
mod token;

#[allow(unused_imports)]
pub use color::{Color, InvalidColorError};
#[allow(unused_imports)]
pub use decoration::{Decoration, InvalidDecorationError};
pub use style::{CurrentStyle, Style};
pub use token::{Token, Tokens};
