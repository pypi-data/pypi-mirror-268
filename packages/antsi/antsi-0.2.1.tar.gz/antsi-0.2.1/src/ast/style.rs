use super::{Color, Decoration};
use indexmap::IndexSet;

/// Styles that can be applied to a piece of text
#[derive(Clone, Debug, Default)]
#[cfg_attr(test, derive(Eq, PartialEq))]
pub struct Style {
    /// The foreground color
    pub foreground: Option<Color>,
    /// The background color
    pub background: Option<Color>,
    /// Additional text decoration (i.e. bold, italic, underline, etc.)
    pub decoration: Option<IndexSet<Decoration>>,
}

impl Style {
    /// Check if the style has any properties
    fn is_empty(&self) -> bool {
        let has_decorations = match &self.decoration {
            Some(decorations) => decorations.is_empty(),
            None => true,
        };
        self.foreground.is_none() && self.background.is_none() && has_decorations
    }

    /// Apply the current style to the text
    pub fn apply(&self, parent: &CurrentStyle, output: &mut String) {
        if self.is_empty() {
            return;
        }

        // typically we'll only have a foreground and single decoration
        let mut codes = Vec::with_capacity(2);

        if let Some(foreground) = self.foreground {
            if foreground != parent.foreground {
                codes.push(foreground.foreground_code())
            }
        }

        if let Some(background) = self.background {
            if background != parent.background {
                codes.push(background.background_code());
            }
        }

        if let Some(decorations) = &self.decoration {
            codes.extend(
                decorations
                    .difference(&parent.decoration)
                    .map(Decoration::apply_code),
            );
        }

        self.append_codes(codes, output);
    }

    /// Reset the style to what it was previously
    pub fn reset(&self, parent: &CurrentStyle, output: &mut String) {
        if self.is_empty() {
            return;
        }

        // typically we'll only have a foreground and single decoration
        let mut codes = Vec::with_capacity(2);

        if let Some(foreground) = self.foreground {
            if foreground != parent.foreground {
                codes.push(parent.foreground.foreground_code());
            }
        }

        if let Some(background) = self.background {
            if background != parent.background {
                codes.push(parent.background.background_code());
            }
        }

        if let Some(decorations) = &self.decoration {
            codes.extend(
                decorations
                    .difference(&parent.decoration)
                    .map(Decoration::remove_code),
            );
        }

        self.append_codes(codes, output);
    }

    /// Append the ANSI codes to the output
    fn append_codes(&self, codes: Vec<&str>, output: &mut String) {
        if codes.is_empty() {
            return;
        }

        output.push_str("\x1b[");
        output.push_str(&codes.join(";"));
        output.push('m');
    }
}

/// The current styles applied to a piece of text
#[derive(Clone, Debug, Default)]
pub struct CurrentStyle {
    foreground: Color,
    background: Color,
    decoration: IndexSet<Decoration>,
}

impl CurrentStyle {
    /// Extend the current style with additional styles from a token
    pub fn extend(&self, style: &Style) -> CurrentStyle {
        let mut current = CurrentStyle::clone(self);

        current.foreground = style.foreground.unwrap_or(current.foreground);
        current.background = style.background.unwrap_or(current.background);
        if let Some(decoration) = &style.decoration {
            current.decoration.extend(decoration.iter());
        }

        current
    }
}

impl From<Style> for CurrentStyle {
    fn from(style: Style) -> Self {
        CurrentStyle {
            foreground: style.foreground.unwrap_or_default(),
            background: style.background.unwrap_or_default(),
            decoration: style.decoration.unwrap_or_default(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{CurrentStyle, Style};

    #[test]
    fn default_is_empty() {
        let style = style!();
        assert!(style.is_empty());
    }

    #[test]
    fn is_empty_with_foreground() {
        let style = style!(fg: Red;);
        assert!(!style.is_empty());
    }

    #[test]
    fn is_empty_with_background() {
        let style = style!(bg: Blue;);
        assert!(!style.is_empty());
    }

    #[test]
    fn is_empty_with_single_decoration() {
        let style = style!(deco: Bold;);
        assert!(!style.is_empty());
    }

    #[test]
    fn is_empty_with_multiple_decorations() {
        let style = style!(deco: Bold, Italic;);
        assert!(!style.is_empty());
    }

    /// Create a sequence of tests
    macro_rules! simple_tests {
    (
        for $function:ident;
        $( $name:ident: $style:expr, $parent:expr => $output:expr ),* $(,)?
    ) => {
            $(
                #[test]
                fn $name () {
                    assert_eq!($function(&$style, $parent), $output);
                }
            )*
        };
    }

    fn apply(style: &Style, parent: Style) -> String {
        let mut output = String::new();
        style.apply(&parent.into(), &mut output);
        output
    }

    #[test]
    fn apply_appends_to_end_of_output() {
        let mut output = String::from("existing");

        let style = style!(fg: Red;);
        style.apply(&CurrentStyle::default(), &mut output);

        assert_eq!(output, "existing\x1b[31m");
    }

    simple_tests! {
        for apply;

        apply_foreground_different_from_parent: style!(fg: Red;), style!(fg: Blue;) => "\x1b[31m",
        apply_foreground_identical_to_parent: style!(fg: Red;), style!(fg: Red;) => "",
        apply_background_different_from_parent: style!(bg: Red;), style!(bg: Blue;) => "\x1b[41m",
        apply_background_identical_to_parent: style!(bg: Red;), style!(bg: Red;) => "",
        apply_single_decoration_different_from_parent: style!(deco: Bold;), style!(deco: Dim;) => "\x1b[1m",
        apply_multiple_decoration_different_from_parent: style!(deco: Bold, Italic;), style!(deco: Dim, Hide;) => "\x1b[1;3m",
        apply_single_decoration_identical_to_parent: style!(deco: Bold;), style!(deco: Bold;) => "",
        apply_multiple_decorations_identical_to_parent: style!(deco: Bold, Italic;), style!(deco: Bold, Italic;) => "",

        apply_foreground_and_background_no_parent: style!(fg: Red; bg: Blue;), style!() => "\x1b[31;44m",
        apply_foreground_and_single_decoration_no_parent: style!(fg: Red; deco: Bold;), style!() => "\x1b[31;1m",
        apply_foreground_and_multiple_decorations_no_parent: style!(fg: Red; deco: Bold, Italic;), style!() => "\x1b[31;1;3m",
        apply_background_and_single_decoration_no_parent: style!(bg: Red; deco: Bold;), style!() => "\x1b[41;1m",
        apply_background_and_mulitple_decoration_no_parent: style!(bg: Red; deco: Bold, Italic;), style!() => "\x1b[41;1;3m",

        apply_foreground_and_background_parent_matches_neither: style!(fg: Red; bg: Blue;), style!(deco: Bold;) => "\x1b[31;44m",
        apply_foreground_and_background_parent_matches_foreground: style!(fg: Red; bg: Blue;), style!(fg: Red;) => "\x1b[44m",
        apply_foreground_and_background_parent_matches_background: style!(fg: Red; bg: Blue;), style!(bg: Blue;) => "\x1b[31m",
        apply_foreground_and_background_parent_matches_both: style!(fg: Red; bg: Blue;), style!(fg: Red; bg: Blue;) => "",

        apply_foreground_and_single_decoration_parent_matches_neither: style!(fg: Red; deco: Bold;), style!(bg: Blue;) => "\x1b[31;1m",
        apply_foreground_and_single_decoration_parent_matches_foreground: style!(fg: Red; deco: Bold;), style!(fg: Red;) => "\x1b[1m",
        apply_foreground_and_single_decoration_parent_matches_decoration: style!(fg: Red; deco: Bold;), style!(deco: Bold;) => "\x1b[31m",
        apply_foreground_and_single_decoration_parent_matches_both: style!(fg: Red; deco: Bold;), style!(fg: Red; deco: Bold;) => "",

        apply_foreground_and_multiple_decorations_parent_matches_neither: style!(fg: Red; deco: Bold, Italic;), style!(bg: Blue;) => "\x1b[31;1;3m",
        apply_foreground_and_multiple_decorations_parent_matches_foreground: style!(fg: Red; deco: Bold, Italic;), style!(fg: Red;) => "\x1b[1;3m",
        apply_foreground_and_multiple_decorations_parent_matches_first_decoration: style!(fg: Red; deco: Bold, Italic;), style!(deco: Bold;) => "\x1b[31;3m",
        apply_foreground_and_multiple_decorations_parent_matches_second_decoration: style!(fg: Red; deco: Bold, Italic;), style!(deco: Italic;) => "\x1b[31;1m",
        apply_foreground_and_mulitple_decorations_parent_matches_both_decorations: style!(fg: Red; deco: Bold, Italic;), style!(deco: Bold, Italic;) => "\x1b[31m",

        apply_background_and_single_decoration_parent_matches_neither: style!(bg: Blue; deco: Bold;), style!(fg: Red;) => "\x1b[44;1m",
        apply_background_and_single_decoration_parent_matches_foreground: style!(bg: Blue; deco: Bold;), style!(bg: Blue;) => "\x1b[1m",
        apply_background_and_single_decoration_parent_matches_decoration: style!(bg: Blue; deco: Bold;), style!(deco: Bold;) => "\x1b[44m",
        apply_background_and_single_decoration_parent_matches_both: style!(bg: Blue; deco: Bold;), style!(bg: Blue; deco: Bold;) => "",

        apply_background_and_multiple_decorations_parent_matches_neither: style!(bg: Blue; deco: Bold, Italic;), style!(fg: Red;) => "\x1b[44;1;3m",
        apply_background_and_multiple_decorations_parent_matches_foreground: style!(bg: Blue; deco: Bold, Italic;), style!(bg: Blue;) => "\x1b[1;3m",
        apply_background_and_multiple_decorations_parent_matches_first_decoration: style!(bg: Blue; deco: Bold, Italic;), style!(deco: Bold;) => "\x1b[44;3m",
        apply_background_and_multiple_decorations_parent_matches_second_decoration: style!(bg: Blue; deco: Bold, Italic;), style!(deco: Italic;) => "\x1b[44;1m",
        apply_background_and_mulitple_decorations_parent_matches_both_decorations: style!(bg: Blue; deco: Bold, Italic;), style!(deco: Bold, Italic;) => "\x1b[44m",
    }

    simple_tests! {
        for apply;

        ansi_code_foreground_default: style!(fg: Default;), style!(fg: Red;) => "\x1b[39m",
        ansi_code_background_default: style!(bg: Default;), style!(bg: Red;) => "\x1b[49m",

        ansi_code_foreground_black: style!(fg: Black;), style!() => "\x1b[30m",
        ansi_code_foreground_red: style!(fg: Red;), style!() => "\x1b[31m",
        ansi_code_foreground_green: style!(fg: Green;), style!() => "\x1b[32m",
        ansi_code_foreground_yellow: style!(fg: Yellow;), style!() => "\x1b[33m",
        ansi_code_foreground_blue: style!(fg: Blue;), style!() => "\x1b[34m",
        ansi_code_foreground_magenta: style!(fg: Magenta;), style!() => "\x1b[35m",
        ansi_code_foreground_cyan: style!(fg: Cyan;), style!() => "\x1b[36m",
        ansi_code_foreground_white: style!(fg: White;), style!() => "\x1b[37m",
        ansi_code_foreground_bright_black: style!(fg: BrightBlack;), style!() => "\x1b[90m",
        ansi_code_foreground_bright_red: style!(fg: BrightRed;), style!() => "\x1b[91m",
        ansi_code_foreground_bright_green: style!(fg: BrightGreen;), style!() => "\x1b[92m",
        ansi_code_foreground_bright_yellow: style!(fg: BrightYellow;), style!() => "\x1b[93m",
        ansi_code_foreground_bright_blue: style!(fg: BrightBlue;), style!() => "\x1b[94m",
        ansi_code_foreground_bright_magenta: style!(fg: BrightMagenta;), style!() => "\x1b[95m",
        ansi_code_foreground_bright_cyan: style!(fg: BrightCyan;), style!() => "\x1b[96m",
        ansi_code_foreground_bright_white: style!(fg: BrightWhite;), style!() => "\x1b[97m",

        ansi_code_background_black: style!(bg: Black;), style!() => "\x1b[40m",
        ansi_code_background_red: style!(bg: Red;), style!() => "\x1b[41m",
        ansi_code_background_green: style!(bg: Green;), style!() => "\x1b[42m",
        ansi_code_background_yellow: style!(bg: Yellow;), style!() => "\x1b[43m",
        ansi_code_background_blue: style!(bg: Blue;), style!() => "\x1b[44m",
        ansi_code_background_magenta: style!(bg: Magenta;), style!() => "\x1b[45m",
        ansi_code_background_cyan: style!(bg: Cyan;), style!() => "\x1b[46m",
        ansi_code_background_white: style!(bg: White;), style!() => "\x1b[47m",
        ansi_code_background_bright_black: style!(bg: BrightBlack;), style!() => "\x1b[100m",
        ansi_code_background_bright_red: style!(bg: BrightRed;), style!() => "\x1b[101m",
        ansi_code_background_bright_green: style!(bg: BrightGreen;), style!() => "\x1b[102m",
        ansi_code_background_bright_yellow: style!(bg: BrightYellow;), style!() => "\x1b[103m",
        ansi_code_background_bright_blue: style!(bg: BrightBlue;), style!() => "\x1b[104m",
        ansi_code_background_bright_magenta: style!(bg: BrightMagenta;), style!() => "\x1b[105m",
        ansi_code_background_bright_cyan: style!(bg: BrightCyan;), style!() => "\x1b[106m",
        ansi_code_background_bright_white: style!(bg: BrightWhite;), style!() => "\x1b[107m",

        ansi_code_decoration_bold: style!(deco: Bold;), style!() => "\x1b[1m",
        ansi_code_decoration_dim: style!(deco: Dim;), style!() => "\x1b[2m",
        ansi_code_decoration_italic: style!(deco: Italic;), style!() => "\x1b[3m",
        ansi_code_decoration_underline: style!(deco: Underline;), style!() => "\x1b[4m",
        ansi_code_decoration_slow_blink: style!(deco: SlowBlink;), style!() => "\x1b[5m",
        ansi_code_decoration_fast_blink: style!(deco: FastBlink;), style!() => "\x1b[6m",
        ansi_code_decoration_invert: style!(deco: Invert;), style!() => "\x1b[7m",
        ansi_code_decoration_hide: style!(deco: Hide;), style!() => "\x1b[8m",
        ansi_code_decoration_strike_through: style!(deco: StrikeThrough;), style!() => "\x1b[9m",
    }

    fn reset(style: &Style, parent: Style) -> String {
        let mut output = String::new();
        style.reset(&parent.into(), &mut output);
        output
    }

    #[test]
    fn reset_appends_to_end_of_output() {
        let mut output = String::from("existing");

        let style = style!(fg: Red;);
        style.reset(&CurrentStyle::default(), &mut output);

        assert_eq!(output, "existing\x1b[39m");
    }

    simple_tests! {
        for reset;

        reset_foreground_different_from_parent: style!(fg: Red;), style!(fg: Blue;) => "\x1b[34m",
        reset_foreground_identical_to_parent: style!(fg: Red;), style!(fg: Red;) => "",
        reset_background_different_from_parent: style!(bg: Red;), style!(bg: Blue;) => "\x1b[44m",
        reset_background_identical_to_parent: style!(bg: Red;), style!(bg: Red;) => "",
        reset_single_decoration_different_from_parent: style!(deco: Bold;), style!(deco: Dim;) => "\x1b[22m",
        reset_multiple_decoration_different_from_parent: style!(deco: Bold, Italic;), style!(deco: Dim, Hide;) => "\x1b[22;23m",
        reset_single_decoration_identical_to_parent: style!(deco: Bold;), style!(deco: Bold;) => "",
        reset_multiple_decorations_identical_to_parent: style!(deco: Bold, Italic;), style!(deco: Bold, Italic;) => "",

        reset_foreground_and_background_no_parent: style!(fg: Red; bg: Blue;), style!() => "\x1b[39;49m",
        reset_foreground_and_single_decoration_no_parent: style!(fg: Red; deco: Bold;), style!() => "\x1b[39;22m",
        reset_foreground_and_multiple_decorations_no_parent: style!(fg: Red; deco: Bold, Italic;), style!() => "\x1b[39;22;23m",
        reset_background_and_single_decoration_no_parent: style!(bg: Red; deco: Bold;), style!() => "\x1b[49;22m",
        reset_background_and_mulitple_decoration_no_parent: style!(bg: Red; deco: Bold, Italic;), style!() => "\x1b[49;22;23m",

        reset_foreground_and_background_parent_matches_neither: style!(fg: Red; bg: Blue;), style!(deco: Bold;) => "\x1b[39;49m",
        reset_foreground_and_background_parent_matches_foreground: style!(fg: Red; bg: Blue;), style!(fg: Red;) => "\x1b[49m",
        reset_foreground_and_background_parent_matches_background: style!(fg: Red; bg: Blue;), style!(bg: Blue;) => "\x1b[39m",
        reset_foreground_and_background_parent_matches_both: style!(fg: Red; bg: Blue;), style!(fg: Red; bg: Blue;) => "",

        reset_foreground_and_single_decoration_parent_matches_neither: style!(fg: Red; deco: Bold;), style!(bg: Blue;) => "\x1b[39;22m",
        reset_foreground_and_single_decoration_parent_matches_foreground: style!(fg: Red; deco: Bold;), style!(fg: Red;) => "\x1b[22m",
        reset_foreground_and_single_decoration_parent_matches_decoration: style!(fg: Red; deco: Bold;), style!(deco: Bold;) => "\x1b[39m",
        reset_foreground_and_single_decoration_parent_matches_both: style!(fg: Red; deco: Bold;), style!(fg: Red; deco: Bold;) => "",

        reset_foreground_and_multiple_decorations_parent_matches_neither: style!(fg: Red; deco: Bold, Italic;), style!(bg: Blue;) => "\x1b[39;22;23m",
        reset_foreground_and_multiple_decorations_parent_matches_foreground: style!(fg: Red; deco: Bold, Italic;), style!(fg: Red;) => "\x1b[22;23m",
        reset_foreground_and_multiple_decorations_parent_matches_first_decoration: style!(fg: Red; deco: Bold, Italic;), style!(deco: Bold;) => "\x1b[39;23m",
        reset_foreground_and_multiple_decorations_parent_matches_second_decoration: style!(fg: Red; deco: Bold, Italic;), style!(deco: Italic;) => "\x1b[39;22m",
        reset_foreground_and_mulitple_decorations_parent_matches_both_decorations: style!(fg: Red; deco: Bold, Italic;), style!(deco: Bold, Italic;) => "\x1b[39m",

        reset_background_and_single_decoration_parent_matches_neither: style!(bg: Blue; deco: Bold;), style!(fg: Red;) => "\x1b[49;22m",
        reset_background_and_single_decoration_parent_matches_foreground: style!(bg: Blue; deco: Bold;), style!(bg: Blue;) => "\x1b[22m",
        reset_background_and_single_decoration_parent_matches_decoration: style!(bg: Blue; deco: Bold;), style!(deco: Bold;) => "\x1b[49m",
        reset_background_and_single_decoration_parent_matches_both: style!(bg: Blue; deco: Bold;), style!(bg: Blue; deco: Bold;) => "",

        reset_background_and_multiple_decorations_parent_matches_neither: style!(bg: Blue; deco: Bold, Italic;), style!(fg: Red;) => "\x1b[49;22;23m",
        reset_background_and_multiple_decorations_parent_matches_foreground: style!(bg: Blue; deco: Bold, Italic;), style!(bg: Blue;) => "\x1b[22;23m",
        reset_background_and_multiple_decorations_parent_matches_first_decoration: style!(bg: Blue; deco: Bold, Italic;), style!(deco: Bold;) => "\x1b[49;23m",
        reset_background_and_multiple_decorations_parent_matches_second_decoration: style!(bg: Blue; deco: Bold, Italic;), style!(deco: Italic;) => "\x1b[49;22m",
        reset_background_and_mulitple_decorations_parent_matches_both_decorations: style!(bg: Blue; deco: Bold, Italic;), style!(deco: Bold, Italic;) => "\x1b[49m",
    }
}
