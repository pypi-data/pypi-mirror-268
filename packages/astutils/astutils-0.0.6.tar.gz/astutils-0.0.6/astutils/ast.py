"""Abstract syntax tree nodes."""
# Copyright 2014-2022 by California Institute of Technology
# All rights reserved. Licensed under 3-clause BSD.
#


class Terminal:
    """Nullary symbol."""

    def __init__(
            self,
            value:
                str,
            dtype:
                str='terminal'
            ) -> None:
        try:
            value + 's'
        except TypeError:
            raise TypeError(
                'value must be a string, '
                f'got: {value}')
        self.type = dtype
        self.value = value

    def __hash__(
            self
            ) -> int:
        return id(self)

    def __repr__(
            self
            ) -> str:
        class_name = type(self).__name__
        return (
            f'{class_name}('
                f'{self.value!r}, '
                f'{self.type!r})')

    def __str__(
            self,
            *arg,
            **kw
            ) -> str:
        return self.value

    def __len__(
            self
            ) -> int:
        return 1

    def __eq__(
            self,
            other
            ) -> bool:
        return (
            hasattr(other, 'type') and
            hasattr(other, 'value') and
            self.type == other.type and
            self.value == other.value)

    def flatten(
            self,
            *arg,
            **kw):
        return self.value


class Operator:
    """Operator with arity > 0."""

    def __init__(
            self,
            operator:
                str,
            *operands
            ) -> None:
        try:
            operator + 'a'
        except TypeError:
            raise TypeError(
                'operator must be string, '
                f'got: {operator}')
        self.type = 'operator'
        self.operator = operator
        self.operands = list(operands)

    def __repr__(
            self
            ) -> str:
        class_name = type(self).__name__
        xyz = ', '.join(map(repr, self.operands))
        return (
            f'{class_name}('
                f'{self.operator!r}, '
                f'{xyz})')

    def __str__(
            self
            ) -> str:
        xyz = ' '.join(map(str, self.operands))
        return f'({self.operator} {xyz})'

    def __len__(
            self
            ) -> int:
        return 1 + sum(map(len, self.operands))

    def flatten(
            self,
            *arg,
            **kw):
        csv = ', '.join(
            x.flatten(*arg, **kw)
            for x in self.operands)
        return f'( {self.operator} {csv} )'
