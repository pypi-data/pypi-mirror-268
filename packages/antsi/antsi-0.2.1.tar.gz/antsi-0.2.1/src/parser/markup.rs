use super::{content::content, style::style, Parser};
use crate::ast::Token;

/// Parse a segment of text with styling
pub(crate) fn markup(p: &mut Parser) -> Option<Token> {
    Some(Token::Styled {
        style: style(p)?,
        content: content(p)?.into(),
    })
}

#[cfg(test)]
mod tests {
    use super::{markup, Parser};
    use crate::{
        error::{Error, Reason},
        lexer::SyntaxKind,
    };

    #[test]
    fn foreground_no_content() {
        assert_parse_snapshot!(markup; "[fg:red]()");
    }

    #[test]
    fn background_no_content() {
        assert_parse_snapshot!(markup; "[bg:blue]()");
    }

    #[test]
    fn decoration_single_style_no_content() {
        assert_parse_snapshot!(markup; "[deco:dim]()");
    }

    #[test]
    fn decoration_multiple_styles_no_content() {
        assert_parse_snapshot!(markup; "[deco:dim,italic]()");
    }

    #[test]
    fn lowercase_alphabetic_content() {
        assert_parse_snapshot!(markup; "[fg:red](hello)");
    }

    #[test]
    fn uppercase_alphabetic_content() {
        assert_parse_snapshot!(markup; "[fg:red](HELLO)");
    }

    #[test]
    fn mixed_alphabetic_content() {
        assert_parse_snapshot!(markup; "[fg:red](hElLo)");
    }

    #[test]
    fn numeric_content() {
        assert_parse_snapshot!(markup; "[fg:red](12345)");
    }

    #[test]
    fn special_character_content() {
        assert_parse_snapshot!(markup; "[fg:red](!@#$%^)");
    }

    #[test]
    fn escaped_character_content() {
        assert_parse_snapshot!(markup; "[fg:red](\\(\\[\\]\\))");
    }

    #[test]
    fn nested_token() {
        assert_parse_snapshot!(markup; "[fg:red]([bg:blue](inner))");
    }

    #[test]
    fn nested_token_with_leading_content() {
        assert_parse_snapshot!(markup; "[fg:red](leading [bg:blue](inner))");
    }

    #[test]
    fn nested_token_with_trailing_content() {
        assert_parse_snapshot!(markup; "[fg:red]([bg:blue](inner) trailing)");
    }

    #[test]
    fn nested_token_with_leading_and_trailing_content() {
        assert_parse_snapshot!(markup; "[fg:red](leading [bg:blue](inner) trailing)");
    }

    #[test]
    fn empty_style_specifiers() {
        let mut parser = Parser::new("[](content)");
        assert_eq!(markup(&mut parser), None);
        assert_eq!(
            parser.errors,
            vec![Error {
                span: Some(span!(1..2)),
                at: SyntaxKind::SquareBracketClose,
                reason: Reason::Expected(vec![
                    SyntaxKind::ForegroundSpecifier,
                    SyntaxKind::BackgroundSpecifier,
                    SyntaxKind::DecorationSpecifier
                ])
            }]
        );
    }

    #[test]
    fn missing_closing_square_bracket_on_style_specifiers() {
        let mut parser = Parser::new("[fg:red(content)");
        assert_eq!(markup(&mut parser), None);
        assert_eq!(
            parser.errors,
            vec![Error {
                span: Some(span!(7..8)),
                at: SyntaxKind::ParenthesisOpen,
                reason: Reason::Expected(vec![SyntaxKind::SquareBracketClose])
            }]
        );
    }

    #[test]
    fn missing_open_parenthesis_for_content() {
        let mut parser = Parser::new("[fg:red]content)");
        assert_eq!(markup(&mut parser), None);
        assert_eq!(
            parser.errors,
            vec![Error {
                span: Some(span!(8..15)),
                at: SyntaxKind::Text,
                reason: Reason::Expected(vec![SyntaxKind::ParenthesisOpen])
            }]
        );
    }

    #[test]
    fn missing_close_parenthesis_for_content() {
        let mut parser = Parser::new("[fg:red](content");
        assert_eq!(markup(&mut parser), None);
        assert_eq!(
            parser.errors,
            vec![Error {
                span: None,
                at: SyntaxKind::Eof,
                reason: Reason::Expected(vec![SyntaxKind::ParenthesisClose])
            }]
        );
    }

    #[test]
    fn content_does_not_immediately_follow_style_specifiers() {
        let mut parser = Parser::new("[fg:red] (content)");
        assert_eq!(markup(&mut parser), None);
        assert_eq!(
            parser.errors,
            vec![Error {
                span: Some(span!(8..9)),
                at: SyntaxKind::Whitespace,
                reason: Reason::Expected(vec![SyntaxKind::ParenthesisOpen])
            }]
        );
    }
}
