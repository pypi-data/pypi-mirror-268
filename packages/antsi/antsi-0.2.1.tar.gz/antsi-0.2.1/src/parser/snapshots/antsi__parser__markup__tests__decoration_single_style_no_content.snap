---
source: src/parser/markup.rs
expression: markup(&mut parser)
---
Some(
    Styled {
        content: [],
        style: Style {
            foreground: None,
            background: None,
            decoration: Some(
                {
                    Dim,
                },
            ),
        },
    },
)
